### ISSUE TITLE
find_program() not working with subproject

### ISSUE THREAD

**Issue by telegraphic** [mesonbuild/meson#11414]
**Describe the bug**
Meson is not finding a program that is provided in a subproject:

```
Executing subproject stringify

stringify| Project name: stringify
stringify| Project version: 1.0.0
stringify| C++ compiler for the host machine: c++ (gcc 5.5.0 "c++ (Ubuntu 5.5.0-12ubuntu1~16.04) 5.5.0 20171010")
stringify| C++ linker for the host machine: c++ ld.bfd 2.26.1
stringify| Build targets in project: 1
stringify| Subproject stringify finished.

Fallback to subproject stringify which provides program stringify
Program stringify found: NO
```

**To Reproduce**

Toy example:

`meson.build`
```meson
project('test_find_program', 'cpp', version: '0.0.1')

subproject('stringify')
find_program('stringify')
```

`subprojects/stringify.wrap`
```
[wrap-git]
url = https://github.com/bf-plugins/stringify
revision = main
depth = 1

[provides]
program_names = stringify
stringify_exe = stringify
```

**Expected behavior**

```
Program stringify found: YES!
```

**system parameters**

```
meson 0.62.0
Python 3.9.5
ninja 1.10.2
Ubuntu linux 18.04
```

**Comment by eli-schwartz** [comment]
The wrap file simply tells meson that it can find the program if it executes the subproject. The subproject still needs to define the program... but creating an executable() isn't enough, you need to register it as well.

https://mesonbuild.com/Reference-manual_builtin_meson.html#mesonoverride_find_program

**Comment by telegraphic** [comment]
Thanks, that approach does indeed work.

FWIW this feels at odds to the docs for [find_program](https://mesonbuild.com/Reference-manual_functions.html#find_program). My reading is that should still find the program based on the search order, which is stated as:

1. Program overrides set via [meson.override_find_program()](https://mesonbuild.com/Reference-manual_builtin_meson.html#mesonoverride_find_program)
2. [[provide] sections](https://mesonbuild.com/Wrap-dependency-system-manual.html#provide-section) in subproject wrap files, if [wrap_mode](https://mesonbuild.com/Builtin-options.html#core-options) is set to forcefallback
3. [[binaries] section](https://mesonbuild.com/Machine-files.html#binaries) in your machine files
4. Directories provided using the dirs: kwarg (see below)
5. Project's source tree relative to the current subdir
6. PATH environment variable
7. [[provide] sections](https://mesonbuild.com/Wrap-dependency-system-manual.html#provide-section) in subproject wrap files, if [wrap_mode](https://mesonbuild.com/Builtin-options.html#core-options) is set to anything other than nofallback

Shouldn't steps 2 and 7 don't pick it up?

**Comment by xclaesse** [comment]
I agree it's a bit hidden information, and improving the doc is always welcome, but if you follow the link about the [provide] section, it says:
> With such wrap file, find_program('myprog') will automatically fallback to use the subproject, **assuming it uses meson.override_find_program('myprog')**.`

**Comment by eli-schwartz** [comment]
The difference between 1 and 2 is that the former applies when the program has already been overridden for unrelated reason, while the latter is a discovery method for finding a place that overrides it for you.

> improving the doc is always welcome

100% agree
